File: api/v1/schemas/realtime_watch.py
# -*- coding: utf-8 -*-
"""Realtime watch API schemas."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


DynamicAdviceStatus = Literal[
    "震荡观察",
    "放量突破观察",
    "突破但量能不足",
    "缩量回踩观察",
    "跌破关键位",
    "接近止损",
    "接近目标价",
    "高位回落风险",
    "放量下跌风险",
    "需要AI复核",
]

RealtimeModelStrategy = Literal["auto", "flash", "pro"]


class RealtimeWatchProfile(BaseModel):
    """重点盯盘每股规则配置。"""

    stock_code: str
    stock_name: Optional[str] = None
    market: Optional[str] = None
    enabled: bool = True
    resistance_price: Optional[float] = None
    support_price: Optional[float] = None
    stop_loss_price: Optional[float] = None
    target_price: Optional[float] = None
    volume_ratio_threshold: float = Field(2.0, ge=0.1, le=20)
    change_percent_threshold: float = Field(3.0, ge=0.1, le=20)
    auto_ai_review_enabled: bool = False
    ai_review_cooldown_minutes: int = Field(30, ge=1, le=1440)
    max_daily_ai_reviews: int = Field(3, ge=1, le=3)
    default_skill: Optional[str] = None
    last_ai_review_at: Optional[str] = None
    ai_review_count_date: Optional[str] = None
    ai_review_count: int = 0
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class RealtimeWatchProfileCreateRequest(BaseModel):
    """新增重点盯盘股票。"""

    stock_code: str = Field(..., min_length=1, max_length=32)
    stock_name: Optional[str] = Field(None, max_length=128)
    market: Optional[str] = Field(None, max_length=16)


class RealtimeWatchProfileUpdateRequest(BaseModel):
    """更新重点盯盘规则配置。"""

    stock_name: Optional[str] = Field(None, max_length=128)
    market: Optional[str] = Field(None, max_length=16)
    enabled: bool = True
    resistance_price: Optional[float] = None
    support_price: Optional[float] = None
    stop_loss_price: Optional[float] = None
    target_price: Optional[float] = None
    volume_ratio_threshold: float = Field(2.0, ge=0.1, le=20)
    change_percent_threshold: float = Field(3.0, ge=0.1, le=20)
    auto_ai_review_enabled: bool = False
    ai_review_cooldown_minutes: int = Field(30, ge=1, le=1440)
    max_daily_ai_reviews: int = Field(3, ge=1, le=3)
    default_skill: Optional[str] = Field(None, max_length=64)


class RealtimeQuoteSnapshot(BaseModel):
    """实时盯盘行情快照。"""

    stock_code: str
    stock_name: Optional[str] = None
    current_price: Optional[float] = None
    change: Optional[float] = None
    change_percent: Optional[float] = None
    volume: Optional[float] = None
    amount: Optional[float] = None
    volume_ratio: Optional[float] = None
    turnover_rate: Optional[float] = None
    today_high: Optional[float] = None
    today_low: Optional[float] = None
    yesterday_high: Optional[float] = None
    yesterday_low: Optional[float] = None
    five_day_high: Optional[float] = None
    five_day_low: Optional[float] = None
    twenty_day_high: Optional[float] = None
    ma5: Optional[float] = None
    ma10: Optional[float] = None
    ma20: Optional[float] = None
    source: str = "best_effort"
    quote_time: Optional[str] = None
    stale: bool = False


class DynamicAdvice(BaseModel):
    """行情规则生成的动态建议，不等于 DecisionSignal。"""

    status: DynamicAdviceStatus
    rules: List[str] = Field(default_factory=list)
    severity: Literal["info", "watch", "warning", "danger"] = "info"
    needs_ai_review: bool = False
    message: str = ""


class RealtimeWatchItem(BaseModel):
    profile: RealtimeWatchProfile
    quote: Optional[RealtimeQuoteSnapshot] = None
    dynamic_advice: DynamicAdvice
    latest_ai_signal: Optional[Dict[str, Any]] = None
    conflict_with_ai: bool = False
    conflict_message: Optional[str] = None


class RealtimeWatchSnapshotResponse(BaseModel):
    items: List[RealtimeWatchItem] = Field(default_factory=list)
    refresh_interval_seconds: int = 60
    max_items: int = 5
    message: Optional[str] = None


class RealtimeWatchProfilesResponse(BaseModel):
    profiles: List[RealtimeWatchProfile] = Field(default_factory=list)
    max_items: int = 5


class RealtimeWatchAiReviewRequest(BaseModel):
    """手动 AI 复核请求。"""

    skill_id: Optional[str] = None
    model_strategy: RealtimeModelStrategy = "auto"
    save_as_default: bool = False
    realtime_context: Optional[Dict[str, Any]] = None


class RealtimeWatchAiReviewResponse(BaseModel):
    task_id: str
    stock_code: str
    status: str = "pending"
    message: str
    cooldown_minutes: int = 30

File: apps/dsa-web/src/api/realtimeWatch.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type { DecisionSignalItem } from '../types/decisionSignals';

export type DynamicAdviceStatus =
  | '震荡观察'
  | '放量突破观察'
  | '突破但量能不足'
  | '缩量回踩观察'
  | '跌破关键位'
  | '接近止损'
  | '接近目标价'
  | '高位回落风险'
  | '放量下跌风险'
  | '需要AI复核';

export type RealtimeModelStrategy = 'auto' | 'flash' | 'pro';

export type RealtimeWatchProfile = {
  stockCode: string;
  stockName?: string | null;
  market?: string | null;
  enabled: boolean;
  resistancePrice?: number | null;
  supportPrice?: number | null;
  stopLossPrice?: number | null;
  targetPrice?: number | null;
  volumeRatioThreshold: number;
  changePercentThreshold: number;
  autoAiReviewEnabled: boolean;
  aiReviewCooldownMinutes: number;
  maxDailyAiReviews?: number | null;
  defaultSkill?: string | null;
  lastAiReviewAt?: string | null;
  aiReviewCountDate?: string | null;
  aiReviewCount: number;
  createdAt?: string | null;
  updatedAt?: string | null;
};

export type RealtimeQuoteSnapshot = {
  stockCode: string;
  stockName?: string | null;
  currentPrice?: number | null;
  change?: number | null;
  changePercent?: number | null;
  volume?: number | null;
  amount?: number | null;
  volumeRatio?: number | null;
  turnoverRate?: number | null;
  todayHigh?: number | null;
  todayLow?: number | null;
  yesterdayHigh?: number | null;
  yesterdayLow?: number | null;
  fiveDayHigh?: number | null;
  fiveDayLow?: number | null;
  twentyDayHigh?: number | null;
  ma5?: number | null;
  ma10?: number | null;
  ma20?: number | null;
  source: string;
  quoteTime?: string | null;
  stale: boolean;
};

export type DynamicAdvice = {
  status: DynamicAdviceStatus;
  rules: string[];
  severity: 'info' | 'watch' | 'warning' | 'danger';
  needsAiReview: boolean;
  message: string;
};

export type RealtimeWatchItem = {
  profile: RealtimeWatchProfile;
  quote?: RealtimeQuoteSnapshot | null;
  dynamicAdvice: DynamicAdvice;
  latestAiSignal?: DecisionSignalItem | null;
  conflictWithAi: boolean;
  conflictMessage?: string | null;
};

export type RealtimeWatchSnapshotResponse = {
  items: RealtimeWatchItem[];
  refreshIntervalSeconds: number;
  maxItems: number;
  message?: string | null;
};

export type RealtimeWatchProfileUpdateRequest = {
  stockName?: string | null;
  market?: string | null;
  enabled: boolean;
  resistancePrice?: number | null;
  supportPrice?: number | null;
  stopLossPrice?: number | null;
  targetPrice?: number | null;
  volumeRatioThreshold: number;
  changePercentThreshold: number;
  autoAiReviewEnabled: boolean;
  aiReviewCooldownMinutes: number;
  defaultSkill?: string | null;
};

function toSnakeProfilePayload(payload: RealtimeWatchProfileUpdateRequest): Record<string, unknown> {
  return {
    stock_name: payload.stockName,
    market: payload.market,
    enabled: payload.enabled,
    resistance_price: payload.resistancePrice,
    support_price: payload.supportPrice,
    stop_loss_price: payload.stopLossPrice,
    target_price: payload.targetPrice,
    volume_ratio_threshold: payload.volumeRatioThreshold,
    change_percent_threshold: payload.changePercentThreshold,
    auto_ai_review_enabled: payload.autoAiReviewEnabled,
    ai_review_cooldown_minutes: payload.aiReviewCooldownMinutes,
    default_skill: payload.defaultSkill,
  };
}

export const realtimeWatchApi = {
  async listProfiles(): Promise<{ profiles: RealtimeWatchProfile[]; maxItems: number }> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/realtime-watch/profiles');
    return toCamelCase(response.data);
  },

  async createProfile(stockCode: string, stockName?: string | null): Promise<RealtimeWatchProfile> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/realtime-watch/profiles', {
      stock_code: stockCode,
      stock_name: stockName,
    });
    return toCamelCase(response.data);
  },

  async updateProfile(
    stockCode: string,
    payload: RealtimeWatchProfileUpdateRequest,
  ): Promise<RealtimeWatchProfile> {
    const response = await apiClient.put<Record<string, unknown>>(
      `/api/v1/realtime-watch/profiles/${encodeURIComponent(stockCode)}`,
      toSnakeProfilePayload(payload),
    );
    return toCamelCase(response.data);
  },

  async deleteProfile(stockCode: string): Promise<void> {
    await apiClient.delete(`/api/v1/realtime-watch/profiles/${encodeURIComponent(stockCode)}`);
  },

  async getSnapshot(refreshIntervalSeconds = 60): Promise<RealtimeWatchSnapshotResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/realtime-watch/snapshot', {
      params: { refresh_interval_seconds: refreshIntervalSeconds },
    });
    return toCamelCase(response.data);
  },

  async submitAiReview(
    stockCode: string,
    payload: {
      skillId?: string | null;
      modelStrategy: RealtimeModelStrategy;
      saveAsDefault: boolean;
      realtimeContext?: Record<string, unknown>;
    },
  ): Promise<{ taskId: string; stockCode: string; status: string; message: string; cooldownMinutes: number }> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/realtime-watch/profiles/${encodeURIComponent(stockCode)}/ai-review`,
      {
        skill_id: payload.skillId,
        model_strategy: payload.modelStrategy,
        save_as_default: payload.saveAsDefault,
        realtime_context: payload.realtimeContext,
      },
      { validateStatus: (status) => status === 200 || status === 400 || status === 409 },
    );
    if (response.status >= 400) {
      const detail = response.data?.detail;
      const message = detail && typeof detail === 'object' && 'message' in detail
        ? String((detail as { message?: unknown }).message || '')
        : 'AI复核提交失败';
      throw new Error(message);
    }
    return toCamelCase(response.data);
  },
};
